Make it possible to override persist id

dnsjts opened the pull request:
This pull request enables overriding the persist id from the configuration file and performs name collision check between the initialized LogPipes in order to prevent name collision related issues.

At the moment, this is relevant for the `LogDriver`s, but we think it might be useful in the future, if the other `LogPipe` types could save their data to a persistent storage. (Therefore, this could be implemented for the `LogDriver`, but this seems more future-proof.)

Fixes #661.
Workarounds #655.

--
TODOs:
* [ ] Rethink the allocation strategy within persist_name implementations
* [ ] Handle the persist_id != NULL case globally
* [x] Do not change the original format in afsocket-dest

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/balabit/syslog-ng/665)
<!-- Reviewable:end -->

ihrwein commented on lib/cfg-parser.c:
```diff
@@ -148,6 +148,7 @@ static CfgLexerKeyword main_keywords[] = {
   { "dns_cache_expire",   KW_DNS_CACHE_EXPIRE },
   { "dns_cache_expire_failed", KW_DNS_CACHE_EXPIRE_FAILED },
   { "pass_unix_credentials", KW_PASS_UNIX_CREDENTIALS },
+  { "persist_name",            KW_PERSIST_ID},
```
`persist_name` -> `persist_id`

dnsjts replied:
Fixed in b01cf912b54096a9307507.

ihrwein commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
+        {
+          if (g_hash_table_lookup(pipe_persist_names, current_pipe->persist_id))
+            {
+              msg_error("Error checking the uniqueness of the persist names, please override it with persist-id option. Shutting down.",
```
`names` -> `ids`

dnsjts replied:
Fixed in b01cf912b54096a9307507.

ihrwein commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
+static gboolean
+has_initialized_pipes_unique_persist_id(const GPtrArray * const initialized_pipes)
+{
+  GHashTable * pipe_persist_names = g_hash_table_new(g_str_hash, g_str_equal);
```
pipe_persist_names -> pipe_persist_ids

dnsjts replied:
Fixed in b01cf912b54096a9307507.

ihrwein commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
 }
 
+static gboolean
+has_initialized_pipes_unique_persist_id(const GPtrArray * const initialized_pipes)
```
this name is a little bit strange to me.

dnsjts replied:
Fixed in b01cf912b54096a9307507.

ihrwein commented on lib/cfg-grammar.y:
```diff
@@ -1104,6 +1108,9 @@ source_reader_option_flags
 	;
 
+driver_option: KW_PERSIST_ID '(' string ')' { log_pipe_set_persist_id(&last_driver->super, g_strdup($3)); free($3); }
+             ;
```
indentation of ;

dnsjts replied:
Fixed in 29e36ae58feff8b92555591603336e82f10c7690.

ihrwein commented on lib/cfg-parser.c:
```diff
@@ -148,6 +148,7 @@ static CfgLexerKeyword main_keywords[] = {
   { "dns_cache_expire",   KW_DNS_CACHE_EXPIRE },
   { "dns_cache_expire_failed", KW_DNS_CACHE_EXPIRE_FAILED },
   { "pass_unix_credentials", KW_PASS_UNIX_CREDENTIALS },
+  { "persist_id",            KW_PERSIST_ID},
```
missing space after KW_PERSIST_ID

dnsjts replied:
Fixed in c2facdd6043e7e1b71d05ba9f9450eaf0d96300c.

ihrwein commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
 }
 
+static gboolean
+are_the_persist_ids_unique(const GPtrArray * const initialized_pipes)
```
much better function name :+1: I'd put a `_` before the name as in other places where a function is `static`.

dnsjts replied:
Generally I would agree with you, but in this file this kind of convention would be uncommon.

ihrwein commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
+are_the_persist_ids_unique(const GPtrArray * const initialized_pipes)
+{
+  GHashTable * pipe_persist_ids = g_hash_table_new(g_str_hash, g_str_equal);
+  gboolean     result           = TRUE;
```
Please don't use vertical alignments. It may be your preference but we have to use the coding convention of the software we are working on.

dnsjts replied:
Fixed in c2cf0bf847581f3d977c977ed2f38100ff73a71f.

ihrwein commented on lib/logpipe.h:
```diff
@@ -363,6 +366,16 @@ log_pipe_append(LogPipe *s, LogPipe *next)
+log_pipe_generate_persist_id(LogPipe * const self)
+{
+  if (self->persist_id != NULL && self->generate_persist_id != NULL)
+    self->persist_id = self->generate_persist_id(self);
```
`generate_persist_id()` returns a different kind of `const` than the `persist_id` field

dnsjts replied:
The type of the return value and the type of the field are the same (`const gchar *` and `gchar const *`), however, mixing the const related conventions is rarely a good idea.

Fixed in cc3bf16146493ed2e8ceb516d0fdab08763265f3.

ihrwein replied:
Ok.

ihrwein commented:
Please fix the style and commit related issues. Nevertheless, this is a very nice PR :)

lbudai commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
+  GHashTable * pipe_persist_ids = g_hash_table_new(g_str_hash, g_str_equal);
+  gboolean result = TRUE;
+
+  for (gint i = 0; i < initialized_pipes->len; ++i)
```
as I remember, I mentioned the foreach function
( https://developer.gnome.org/glib/stable/glib-Pointer-Arrays.html#g-ptr-array-foreach )

dnsjts replied:
It would be nice, but the function has `void` return type, and I would need a non-void type in order to know if was there a name collision or not.
This could be worked around with an output parameter, but that would make the code more complex.

lbudai replied:
ok, got it

dnsjts commented:
Rebased to the master, `git diff 8bfb8e2 cc3bf16` and `git diff 5205a76 2669795` should be the same.

ihrwein commented:
please rename `persist_name` to `persist_id` in the functions.

ihrwein commented:
and it's not `persist_name_function` but `generate_persist_id`

gregory094 commented:
If you correct that one note of mine I'll give my :+1: .

dnsjts commented:
@lbudai , the commits have been fixed up, and are ready to review. Unfortunately, I didn't find better type than the HashTree.

ihrwein commented:
@dnsjts what's the current state of this? Is it ready for merge? If that's so please rebase it again so @lbudai can merge it.

lbudai commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
+{
+  GHashTable * pipe_persist_ids = g_hash_table_new(g_str_hash, g_str_equal);
+  gboolean     result           = TRUE;
+
```
In syslog-ng code base we don't use this kind of vertical alignment.
and we bind the pointer to the variable (` *initialized_pipes` and later, inside the for cycle, ` *current_pipe` is suggested). 
We've talked about this ;-)

krajorama commented on lib/logpipe.h:
```diff
@@ -221,6 +222,8 @@ struct _LogPipe
   gboolean (*init)(LogPipe *self);
   gboolean (*deinit)(LogPipe *self);
 
+  const gchar *(*generate_persist_id)(LogPipe *self);
```
argument should be const LogPipe

krajorama replied:
I'd prefer generate_persist_name, since returned value is usually prefixed by the driver.

krajorama commented on lib/logpipe.h:
```diff
@@ -363,6 +366,12 @@ log_pipe_append(LogPipe *s, LogPipe *next)
+log_pipe_set_persist_id(LogPipe *self, const gchar *persist_id);
+
+const gchar *
+log_pipe_get_persist_id(const LogPipe *self);
```
i'd prefer log_pipe_get_persist_name, since the returned value is usually prefix+persist_id (e.g. afsocket_dd.persistid)

krajorama commented on modules/affile/affile-source.c:
```diff
@@ -456,6 +456,7 @@ affile_sd_new_instance(gchar *filename, GlobalConfig *cfg)
   self->super.super.super.deinit = affile_sd_deinit;
   self->super.super.super.notify = affile_sd_notify;
   self->super.super.super.free_fn = affile_sd_free;
+  self->super.super.super.generate_persist_id = affile_sd_format_persist_name;
```
We'll need a wrapper around affile_sd_format_persist_name to accept const LogPipe* and turned that into the real "class". Now this code and the other similar give warning of incompatible type.

krajorama commented on modules/afsocket/afsocket-dest.c:
```diff
@@ -83,18 +83,60 @@ afsocket_dd_set_keep_alive(LogDriver *s, gboolean enable)
+    }
+  else
+    {
+      gchar buf[64];
```
64 -> MAX_SOCKADDR_STRING

krajorama commented on modules/afsocket/afsocket-source.c:
```diff
@@ -252,19 +252,60 @@ afsocket_sd_set_max_connections(LogDriver *s, gint max_connections)
+static inline gchar *
+afsocket_sd_format_persist_name(const AFSocketSourceDriver *self, gboolean listener_name)
+{
+  return listener_name ? afsocket_sd_format_listener_name : afsocket_sd_format_connections_name;
```
Missing parameter (self) when calling the functions. Probably needs to return const gchar*

krajorama commented on lib/logpipe.c:
```diff
@@ -94,3 +95,21 @@ log_pipe_forward_notify(LogPipe *self, gint notify_code, gpointer user_data)
+log_pipe_set_persist_id(LogPipe *self, const gchar *persist_id)
+{
+  if (self->persist_id != NULL)
+    g_free(self->persist_id);
```
g_free((gchar*)self->persist_id)  to avoid warning

bazsi replied:
and g_free handles NULL just fine, no need to check

dnsjts replied:
Fixed in 82cab03647aa464ca55c19072e381bc5e71268f7.

krajorama commented on lib/logpipe.c:
```diff
@@ -94,3 +95,21 @@ log_pipe_forward_notify(LogPipe *self, gint notify_code, gpointer user_data)
+}
+
+const gchar *
+log_pipe_get_persist_id(const LogPipe *self)
```
I'd prefer log_pipe_get_persist_name (see at declaration)

bazsi replied:
I agree

krajorama commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
 }
 
+static gboolean
+verify_unique_persist_ids_among_pipes(const GPtrArray *initialized_pipes)
```
I'd prefer to replace ids , id with names, name in all places, see logpipe.h comments.

bazsi replied:
and please prefix the function with an underscore to indicate that it is a private.

krajorama commented:
In general: the return value of persist name/id generator functions can be const. The functions in cfg.c (cfg_persist_config_fetch, cfg_persist_config_add) should accept const gchar* name, instead of gchar* name.

bazsi commented on lib/cfg-parser.c:
```diff
@@ -148,6 +148,7 @@ static CfgLexerKeyword main_keywords[] = {
   { "dns_cache_expire_failed", KW_DNS_CACHE_EXPIRE_FAILED },
   { "pass_unix_credentials", KW_PASS_UNIX_CREDENTIALS },
+  { "persist_id",            KW_PERSIST_ID },
 
```
please specify minimum config version where this is recognized to avoid keyword <> identifier clashes.

bazsi replied:
sometimes the option is named persist_id in function names, in other cases as persist_name, we should be consistent about this.

bazsi commented on lib/cfg-tree.c:
```diff
@@ -1065,6 +1066,38 @@ cfg_tree_compile(CfgTree *self)
+              msg_error("Error checking the uniqueness of the persist ids, please override it with persist-id option. Shutting down.",
+                        evt_tag_str("persist_id", current_pipe_name),
+                        NULL);
+              result = FALSE;
```
you can give more detail here, LogPipe instances have location information that can be logged using log_pipe_location_tag()

bazsi replied:
please note that the value returned by log_pipe_get_persist_id() is stored in a static variable that gets overwritten,by subsequent calls of log_pipe_get_persist_id() into the same driver. Thus if you have multiple instances of the same driver, the persist name stored in the hash would be overwritten.

It would probably make sense to change the allocation strategy within persist_name implementations now as they became a public API. Perhaps it would even make sense to move the persist_name to be stored in the LogPipe instance as member and automatically generated _once_ at initialization and then just queried. That would solve the storage question as well.

bazsi commented on lib/cfg-grammar.y:
```diff
@@ -1103,6 +1107,10 @@ source_reader_option_flags
+    : KW_PERSIST_ID '(' string ')' { log_pipe_set_persist_id(&last_driver->super, g_strdup($3)); free($3); }
+    ;
+
 threaded_dest_driver_option
```
please take care about whitespace here. bison files are already pretty unreadable, using a compressed form like this makes that more difficult.

dnsjts replied:
The indentation character is already mixed, I would clarify this in a new pull request.

bazsi commented on lib/logpipe.c:
```diff
@@ -94,3 +95,21 @@ log_pipe_forward_notify(LogPipe *self, gint notify_code, gpointer user_data)
+    return self->generate_persist_id(self);
+  else
+    return self->persist_id;
+}
```
I would love to see the common case handled by common code, e.g. when persist_id is in fact specified in the configuration, it would be nice _NOT_ to call the LogPipe specific function. Now all of them has to handle the persist_id != NULL case explicitly, which is very easy to get wrong. If the user specified persist_id it should be used without the potential of messing it up in the driver code.

bazsi commented on modules/afsocket/afsocket-dest.c:
```diff
@@ -83,18 +83,60 @@ afsocket_dd_set_keep_alive(LogDriver *s, gboolean enable)
+{
+  return qfile ? afsocket_dd_format_qfile_name(self) : afsocket_dd_format_connections_name(self);
+}
+
```
We should probably drop the qfile boolean argument and this function entirely. call sites should be calling the specific format_name function instead.

dnsjts replied:
Fixed in 601bafce3d795201807e926fac73a374c9323c1d.

bazsi commented on modules/afsocket/afsocket-dest.c:
```diff
@@ -83,18 +83,60 @@ afsocket_dd_set_keep_alive(LogDriver *s, gboolean enable)
+             "%s.connections",
+             afsocket_dd_format_name(self));
+
   return persist_name;
```
I am not sure this doesn't change the persist_name value at the end. That shouldn't be changed lightly, as these names are persisted into the persist file. And changing the name might cause current file positions to be changed across reloads.

dnsjts commented:
persist_id -> persist_name is in progress.

@bazsi, could you review 56c33b4085fa82d31b69257dc93efaef7d4cac1a, because this transformation simplifies a lot of setters, but I fail to find any of its drawbacks, and I suspect it has.

dnsjts commented:
Review status: 0 of 18 files reviewed at latest revision, 8 unresolved discussions.

---

*[lib/cfg-parser.c, line 152 \[r4\]](https://reviewable.io:443/reviews/balabit/syslog-ng/665#-K1Zuhrm2KrWqouH4LdT-r4-152:-K3xrD3NcUd5BWKkZEuw:1841519536) ([raw file](https://github.com/balabit/syslog-ng/blob/4137cb5fccc4cbd8d63582f8d7e67d178bcbf507/lib/cfg-parser.c#L152)):*
The minimal config version has been specified and all persist_id have been renamed to persist_name.

---

*[lib/cfg-tree.c, line 1070 \[r3\]](https://reviewable.io:443/reviews/balabit/syslog-ng/665#-K1Zuhrm2KrWqouH4LdU-r3-1070:-K3xtoyXe5L2wiA1ZmcV:-74546224) ([raw file](https://github.com/balabit/syslog-ng/blob/9ff7623e2005e85a38fdb5f6b00cfcf4ed1a16b6/lib/cfg-tree.c#L1070)):*
id-s have been renamed to name-s, underscore has been added.

---

*[lib/cfg-tree.c, line 1087 \[r4\]](https://reviewable.io:443/reviews/balabit/syslog-ng/665#-K1Zuhrm2KrWqouH4LdU-r4-1087:-K3xyecyjx2n-L3vd0zH:-1400072318) ([raw file](https://github.com/balabit/syslog-ng/blob/4137cb5fccc4cbd8d63582f8d7e67d178bcbf507/lib/cfg-tree.c#L1087)):*
* log_pipe_location_tag has been added
* rethinking the allocation strategy has been added to a todo list

---

*[lib/logpipe.c, line 109 \[r3\]](https://reviewable.io:443/reviews/balabit/syslog-ng/665#-K1Zuhrm2KrWqouH4LdV-r3-109:-K3xzVRLhSbtwCdiGPs6:1755713581) ([raw file](https://github.com/balabit/syslog-ng/blob/9ff7623e2005e85a38fdb5f6b00cfcf4ed1a16b6/lib/logpipe.c#L109)):*
Renamed.

---

*[lib/logpipe.h, line 225 \[r3\]](https://reviewable.io:443/reviews/balabit/syslog-ng/665#-K1Zuhrm2KrWqouH4LdW-r3-225:-K3xzdoZPqtUVnmXQARN:1056155171) ([raw file](https://github.com/balabit/syslog-ng/blob/9ff7623e2005e85a38fdb5f6b00cfcf4ed1a16b6/lib/logpipe.h#L225)):*
* const qualifier has been added
* function pointer has been renamed

---

*[lib/logpipe.h, line 373 \[r3\]](https://reviewable.io:443/reviews/balabit/syslog-ng/665#-K1Zuhrm2KrWqouH4LdW-r3-373:-K3xzmoTtVThqgZiKn45:1755713581) ([raw file](https://github.com/balabit/syslog-ng/blob/9ff7623e2005e85a38fdb5f6b00cfcf4ed1a16b6/lib/logpipe.h#L373)):*
Renamed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/balabit/syslog-ng/665)*
<!-- Sent from Reviewable.io -->

lbudai commented:
@dnsjts : is it  ready for merge at your side ?

dnsjts commented:
IHMO the original task is completed, but other issues were mentioned.
I would consider these out-of-scope issues, and the solution for these might not be trivial.

However, the decision should not be mine.

lbudai commented:
@dnsjts : any update?

dnsjts commented:
It is in progress, I'm currently resolving the issues introduced by the `-Werror` flag, the travis build should be green soon.

ihrwein commented:
I think it's good enough: +1

But, we may need to eliminate the common `if (self->persist_name) {} else {}` pattern in the code in another PR.

lbudai commented on lib/logthrdestdrv.c:
```diff
@@ -32,8 +32,8 @@ log_threaded_dest_driver_format_seqnum_for_persist(LogThrDestDriver *self)
 {
   static gchar persist_name[256];
```
note : maybe it is time to using a `GString` instead of the static buffer?
(or, at least we should tell the user that the persist name length has a max length...).

bazsi replied:
if it's only an sprintf, then I would use g_strdup_printf(), but I guess
the problem is allocation and how to free this allocation.

-- 
Bazsi

On Wed, Jul 13, 2016 at 1:10 PM, Budai Laszlo <notifications@github.com>
wrote:

> In lib/logthrdestdrv.c
> <https://github.com/balabit/syslog-ng/pull/665#discussion_r70605993>:
>
> > @@ -32,8 +32,8 @@ log_threaded_dest_driver_format_seqnum_for_persist(LogThrDestDriver *self)
> >  {
> >    static gchar persist_name[256];
>
> note : maybe it is time to using a GString instead of the static buffer?
> (or, at least we should tell the user that the persist name length has a
> max length...).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/balabit/syslog-ng/pull/665/files/02536ffae49a761b9e76709963203a2eb11057b0..3571c7fda90bface84257cd23df7a87c47b455c9#r70605993>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AArlduEEbuzQmB-qe_ib2htXR1YDzaaxks5qVMecgaJpZM4F2VRA>
> .
>

lbudai commented:
@ihrwein : yeah, it is good enough, and agree: the mentioned fixes could be resolved in a separate PR.

lbudai commented:
@dnsjts : Thanks!
